src/ir/effects.h

```diff
@@ -446,14 +447,19 @@ struct EffectAnalyzer
           implicitTrap = true;
           break;
         }
-        default: {}
+        default: {
+        }
       }
     }
   }
   void visitSelect(Select* curr) {}
   void visitDrop(Drop* curr) {}
   void visitReturn(Return* curr) { branchesOut = true; }
-  void visitHost(Host* curr) {
+  void visitMemorySize(MemorySize* curr) {
+    // Atomics are sequentially consistent with memory.size.
+    isAtomic = true;
```

Comment:
Dangerous half-knowledge.

Reply:
Also, if the case, eliminating unused `memory.size` instructions might not actually be a good idea. Hmm.

Reply:
Yes, this is true (see figure 10 in the [paper](https://dl.acm.org/doi/pdf/10.1145/3360559)). When atomics are enabled and the memory is shared, we could replace unused `memory.size` instructions with `atomic.fence` instructions, which have the same ordering guarantees but do no extra work. If atomics are not enabled or the memory is not shared, we can continue removing unused `memory.size`s. Where is the code that removes these?

These should also be modeled as reading memory, since it reads the size of memory and we don't have a more fine grained effect for making that distinction.

Reply:
Thanks, makes sense :)

> Where is the code that removes these?

The code location in question is in `Vacuum::optimize`, as a side-effect of adding the new MemorySizeId case to the ConstId case [here](https://github.com/WebAssembly/binaryen/pull/3137/files#diff-866573bb9b1d1cab558ac4e056b0615eR115). Going to move the case to the always needed case instead.